types/blob/blob.go

```diff
@@ -0,0 +1,150 @@
+	NMTIgnoreMaxNamespace = true
+)
+
+var (
+	ErrBlobNotFound = errors.New("blob: not found")
+	ErrInvalidProof = errors.New("blob: invalid proof")
+)
+
+// Proof is a collection of nmt.Proofs that verifies the inclusion of the data.
+type Proof []*nmt.Proof
+
+func (p Proof) Len() int { return len(p) }
+
+// equal is a temporary method that compares two proofs.
+// should be removed in BlobService V1.
+func (p Proof) equal(input Proof) error {
+	if p.Len() != input.Len() {
+		return ErrInvalidProof
+	}
+
+	for i, proof := range p {
+		pNodes := proof.Nodes()
+		inputNodes := input[i].Nodes()
+		for i, node := range pNodes {
+			if !bytes.Equal(node, inputNodes[i]) {
+				return ErrInvalidProof
+			}
+		}
+
+		if proof.Start() != input[i].Start() || proof.End() != input[i].End() {
+			return ErrInvalidProof
+		}
+
+		if !bytes.Equal(proof.LeafHash(), input[i].LeafHash()) {
+			return ErrInvalidProof
+		}
+
+	}
+	return nil
+}
+
+type jsonProof struct {
+	Start int      `json:"start"`
+	End   int      `json:"end"`
+	Nodes [][]byte `json:"nodes"`
+}
+
+func (p *Proof) MarshalJSON() ([]byte, error) {
+	proofs := make([]jsonProof, 0, p.Len())
+	for _, pp := range *p {
+		proofs = append(proofs, jsonProof{
+			Start: pp.Start(),
+			End:   pp.End(),
+			Nodes: pp.Nodes(),
+		})
+	}
+
+	return json.Marshal(proofs)
+}
+
+func (p *Proof) UnmarshalJSON(data []byte) error {
+	var proofs []jsonProof
+	err := json.Unmarshal(data, &proofs)
+	if err != nil {
+		return err
+	}
+
+	nmtProofs := make([]*nmt.Proof, len(proofs))
+	for i, jProof := range proofs {
+		nmtProof := nmt.NewInclusionProof(jProof.Start, jProof.End, jProof.Nodes, NMTIgnoreMaxNamespace) // ipld.NMTIgnoreMaxNamespace
+		nmtProofs[i] = &nmtProof
+	}
+
+	*p = nmtProofs
+	return nil
+}
+
+// Commitment is a Merkle Root of the subtree built from shares of the Blob.
+// It is computed by splitting the blob into shares and building the Merkle subtree to be included
+// after Submit.
+type Commitment []byte
+
+func (com Commitment) String() string {
+	return string(com)
+}
+
+// Equal ensures that commitments are the same
+func (com Commitment) Equal(c Commitment) bool {
+	return bytes.Equal(com, c)
+}
+
+// Blob represents any application-specific binary data that anyone can submit to Celestia.
+type Blob struct {
+	rpcproto.Blob `json:"blob"`
+
+	Commitment Commitment `json:"commitment"`
+}
+
+// Namespace returns blob's namespace.
+func (b *Blob) Namespace() namespace.ID {
+	return append([]byte{uint8(b.NamespaceVersion)}, b.NamespaceId...)
+}
+
+type jsonBlob struct {
+	Namespace    namespace.ID `json:"namespace"`
+	Data         []byte       `json:"data"`
+	ShareVersion uint32       `json:"share_version"`
+	Commitment   Commitment   `json:"commitment"`
+}
+
+func (b *Blob) MarshalJSON() ([]byte, error) {
+	blob := &jsonBlob{
+		Namespace:    b.Namespace(),
+		Data:         b.Data,
+		ShareVersion: b.ShareVersion,
+		Commitment:   b.Commitment,
```

Comment:
i think you can define Blob proto type that has the commitment (repeated bytes) in the proto itself to avoid having to define another wrapper type. celestia-node did the wrapper because they were reusing the celestia-app defined Blob proto. unless you have a reason for having Blob proto type and non-proto Commitment type.

Reply:
No, reason other than keeping in sync with node; will update 👍🏽

Reply:
marshal/unmarshal are expensive, hence if we have everything done once (as part of proto) it is better, instead of doing it twice via wrapped type. thanks.

Reply:
Done.